src/lib/cpctplusdyndist.rs

```diff
@@ -0,0 +1,1088 @@
+                              .map(|rprs| {
+                                   let (la_idx, pstack) = apply_repairs(self.parser,
+                                                                        in_la_idx,
+                                                                        start_pstack.clone(),
+                                                                        &mut None,
+                                                                        &rprs);
+                                   (pstack, la_idx, rprs)
+                               })
+                              .collect::<Vec<(Cactus<StIdx>, usize, Vec<ParseRepair>)>>();
+
+        // First try parsing each candidate repair until it hits an error or exceeds TRY_PARSE_AT_MOST
+        // lexemes.
+
+        let mut todo = Vec::new();
+        todo.resize(cnds.len(), true);
+        let mut remng = cnds.len(); // Remaining items in todo
+        let mut i = 0;
+        'b: while i < TRY_PARSE_AT_MOST && remng > 1 {
+            let mut j = 0;
+            while j < todo.len() {
+                if Instant::now() >= finish_by {
+                    break 'b;
+                }
+
+                if !todo[j] {
+                    j += 1;
+                    continue;
+                }
+                let cnd = &mut cnds[j];
+                if cnd.1 > in_la_idx + i {
+                    j += 1;
+                    continue;
+                }
+                let (new_la_idx, new_pstack) = self.parser.lr_cactus(None,
+                                                                     in_la_idx + i,
+                                                                     in_la_idx + i + 1,
+                                                                     cnd.0.clone(),
+                                                                     &mut None);
+                if new_la_idx == in_la_idx + i {
+                    todo[j] = false;
+                    remng -= 1;
+                } else {
+                    cnd.0 = new_pstack;
+                    cnd.1 += 1;
+                }
+                j += 1;
+            }
+            i += 1;
+        }
+
+        // Now rank the candidates into descending order, first by how far they are able to parse, then
+        // by the number of actions in the repairs (the latter is somewhat arbitrary, but matches the
+        // intuition that "repairs which affect the shortest part of the string are preferable").
+        cnds.sort_unstable_by(|x, y| {
+            y.1.cmp(&x.1)
+               .then_with(|| x.2.len()
+                                .cmp(&y.2.len()))
+               .then_with(|| x.2.cmp(&y.2))
+        });
+
+        cnds.into_iter()
+            .map(|x| x.2)
+            .collect::<Vec<Vec<ParseRepair>>>()
+    }
+
+    /// Return the distance from `st_idx` at input position `la_idx`, given the current `repairs`.
+    /// Returns `None` if no route can be found.
+    fn dyn_dist(&self,
```

Comment:
Is this the same function as in mf.rs? Is it worth it to move this somewhere else so it can be shared (esp. if there are other shared functions as well)?

Reply:
It is sort-of shared, but it's relative to `self`, so it's hard to do without lots of refactoring :/ In the long run, I expect to delet `cpctplus.rs` and `cpctplusdyndist.rs`, so I'm not sure it's worth the effort?

Reply:
Fair enough.